include/swift/Basic/SupplementaryOutputPaths.h

```diff
@@ -21,49 +21,106 @@
+  ///
   /// Currently only makes sense when the compiler has whole module knowledge.
   /// The modes for which it makes sense incuide both WMO and the "merge
   /// modules" job that happens after the normal compilation jobs. That's where
   /// the header is emitted in single-file mode, since it needs whole-module
   /// information.
-
+  ///
+  /// \sa swift::printAsObjC
   std::string ObjCHeaderOutputPath;
 
   /// The path to which we should emit a serialized module.
   /// It is valid whenever there are any inputs.
+  ///
+  /// This binary format is used to describe the interface of a module when
+  /// imported by client source code. The swiftmodule format is described in
+  /// docs/Serialization.rst.
+  ///
+  /// \sa swift::serialize
   std::string ModuleOutputPath;
 
   /// The path to which we should emit a module documentation file.
   /// It is valid whenever there are any inputs.
+  ///
+  /// This binary format stores doc comments and other information about the
+  /// declarations in a module.
+  ///
+  /// \sa swift::serialize
   std::string ModuleDocOutputPath;
 
   /// The path to which we should output a Make-style dependencies file.
   /// It is valid whenever there are any inputs.
+  ///
+  /// Swift's compilation model means that Make-style dependencies aren't
+  /// well-suited to model fine-grained dependencies. See docs/Driver.md for
+  /// more information.
+  ///
+  /// \sa ReferenceDependenciesFilePath
   std::string DependenciesFilePath;
 
-  /// The path to which we should output a Swift reference dependencies file.
+  /// The path to which we should output a Swift "reference dependencies" file.
   /// It is valid whenever there are any inputs.
+  ///
+  /// "Reference dependencies" track dependencies on a more fine-grained level
+  /// than just "this file depends on that file". With Swift's "implicit
+  /// visibility" within a module, that becomes very important for any sort of
+  /// incremental build. These files are consumed by the Swift driver to decide
+  /// whether a source file needs to be recompiled during a build. See
+  /// docs/DependencyAnalysis.rst for more information.
+  ///
+  /// \sa swift::emitReferenceDependencies
+  /// \sa DependencyGraph
   std::string ReferenceDependenciesFilePath;
 
   /// Path to a file which should contain serialized diagnostics for this
   /// frontend invocation.
+  ///
+  /// This uses the same serialized diagnostics format as Clang, for tools that
+  /// want machine-parseable diagnostics. There's a bit more information on
+  /// how clients might use this in docs/Driver.md.
+  ///
+  /// \sa swift::serialized_diagnostics::createConsumer
   std::string SerializedDiagnosticsPath;
 
   /// The path to which we should output fix-its as source edits.
+  ///
+  /// This is a JSON-based format that is used by the migrator, but is not
+  /// really vetted for anything else.
+  ///
+  /// \sa swift::writeEditsInJson
   std::string FixItsOutputPath;
```

Comment:
@nkcsgexi This format could use some documentation somewhere, even if includes a disclaimer about using it for anything else!